contracts/validator-manager/PoSValidatorManager.sol

```diff
@@ -486,46 +424,61 @@ abstract contract PoSValidatorManager is
         // Set the delegator status to pending removed, so that it can be properly removed in
         // the complete step, even if the delivered nonce is greater than the nonce used to
         // initialize the removal.
-        delegator.status = DelegatorStatus.PendingRemoved;
+        $._delegatorStakes[delegationID].status = DelegatorStatus.PendingRemoved;
 
-        uint64 validatorUptimeSeconds;
-        uint64 delegationEndTime;
         if (validator.status == ValidatorStatus.Active) {
             if (includeUptimeProof) {
                 // Uptime proofs include the absolute number of seconds the validator has been active.
-                validatorUptimeSeconds = _getUptime(validationID, messageIndex);
+                $._completedValidationUptimeSeconds[validationID] =
+                    _getUptime(validationID, messageIndex);
             }
-            uint64 newValidatorWeight = validator.weight - delegator.weight;
-            (delegator.endingNonce,) = _setValidatorWeight(validationID, newValidatorWeight);
 
-            delegationEndTime = uint64(block.timestamp);
+            ($._delegatorStakes[delegationID].endingNonce,) =
+                _setValidatorWeight(validationID, validator.weight - delegator.weight);
+            _calculateDelegationReward(delegationID);
+        } else if (validator.status == ValidatorStatus.Completed) {
+            _calculateDelegationReward(delegationID);
+            return _completeEndDelegation(delegationID);
         } else {
-            // If the validation period has already ended, we have saved the uptime.
-            // Further, it is impossible to retrieve an uptime proof for an already ended validation,
-            // so there's no need to check any uptime proof provided in this function call.
-            validatorUptimeSeconds = $._completedValidationUptimeSeconds[validationID];
+            revert InvalidValidatorStatus();
+        }
+
+        emit DelegatorRemovalInitialized({delegationID: delegationID, validationID: validationID});
+    }
+
+    function _calculateDelegationReward(bytes32 delegationID) private {
```

Comment:
This could just be my personal preference, so feel free to take it or leave it: I think it'd be a bit cleaner to have the `_calculate...` methods be `pure` or `view`, and then leave it to the caller to make any resulting state updates. That way, the `delegator` struct could be passed directly to the helper to avoid the extra look up, and it is clear that caller is responsible for saving any changes.

Reply:
would be my pref too just based on naming of func

Reply:
Yeah, this is a good call :)